### PR TITLE
BUG: Patch rank() uint64 behavior

### PR DESCRIPTION
**Pull request by gfyoung**
Adds `uint64` ranking functions to `algos.pyx` to allow for proper ranking with `uint64`.

Also introduces partial patch for `factorize()` by adding `uint64` hashtables and vectors for
usage. However, this patch is only partial because the larger bug of non-support for `uint64`
in `Index` has not been fixed (**UPDATE**: tackled in #14937):
~~~python
>>> from pandas import Index, np
>>> Index(np.array([2**63], dtype=np.uint64))
Int64Index([-9223372036854775808], dtype='int64')
~~~

Also patches a bug in `UInt64HashTable` from #14915 that had an erroneous null condition that was caught during testing and was hence removed.

Note there is overlap with #14934 with the implementation of `is_signed_integer_dtype` and `is_unsigned_integer_dtype`.  That PR should be merged before this one.

### CODE REVIEW

File: pandas/algos.pyx [hunk 1]
```diff
@@ -104,6 +104,21 @@ cdef _take_2d_int64(ndarray[int64_t, ndim=2] values,
             result[i, j] = values[i, indexer[i, j]]
     return result
 
+cdef _take_2d_uint64(ndarray[uint64_t, ndim=2] values,
```
Comment (jreback) [inline pandas/algos.pyx]: not sure why float64/int64/object are here either, they should also be in algos_take_helper.pxi.in

can you move there (could be in same PR or another ok)

File: pandas/algos.pyx [hunk 2]
```diff
@@ -286,6 +301,80 @@ def rank_1d_int64(object in_arr, ties_method='average', ascending=True,
         return ranks
 
 
+def rank_1d_uint64(object in_arr, ties_method='average', ascending=True,
```
Comment (jreback) [inline pandas/algos.pyx]: same we need to create a algos_rank_helper.pxi.in and move all of the rank routines (and then add uint64). prob easiest to move, then add (in another PR), but up to you.

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -574,7 +579,27 @@ def mode(values):
 def rank(values, axis=0, method='average', na_option='keep',
          ascending=True, pct=False):
     """
+    Rank the values along a given axis.
 
+    Parameters
```
Comment (gfyoung) [inline pandas/core/algorithms.py]: 👍 : can't really patch if you don't know what you're patching 😄

File: pandas/core/algorithms.py [hunk 4]
```diff
@@ -584,6 +609,8 @@ def rank(values, axis=0, method='average', na_option='keep',
         f, values = _get_data_algo(values, _rank2d_functions)
         ranks = f(values, axis=axis, ties_method=method,
                   ascending=ascending, na_option=na_option, pct=pct)
+    else:
+        raise TypeError("Array with ndim > 2 are not supported.")
```
Comment (gfyoung) [inline pandas/core/algorithms.py]: There is now.

File: pandas/algos.pyx (hunk 2 above)
Comment (gfyoung) [inline pandas/algos.pyx]: Indeed.  Done.

File: pandas/algos.pyx (hunk 1 above)
Comment (gfyoung) [inline pandas/algos.pyx]: Makes sense.  Done.

**Comment by gfyoung** [comment]
@jreback : Addressed all comments, and everything is green and passing. Ready to merge if there are no other concerns.

**Comment by gfyoung** [comment]
@jreback : With #14934 merged, this is ready to be merged now.

File: pandas/src/algos_rank_helper.pxi.in [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+{{for name, ctype, pos_nan_value, neg_nan_value in dtypes}}
+
+
+{{if name == 'generic'}}
+
+
+def rank_1d_{{name}}(object in_arr, bint retry=1, ties_method='average',
+                     ascending=True, na_option='keep', pct=False):
+{{else}}
+
```
Comment (jreback) [inline pandas/src/algos_rank_helper.pxi.in]: prob could just have the same signature for object (and just ignore that arg no?)

File: pandas/src/algos_rank_helper.pxi.in [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+            raise
+
+        valid_locs = (~mask).nonzero()[0]
+        ranks.put(valid_locs, rank_1d_generic(values.take(valid_locs), 0,
+                                              ties_method=ties_method,
+                                              ascending=ascending))
+        np.putmask(ranks, mask, np.nan)
+        return ranks
+    {{else}}
+    # py2.5/win32 hack, can't pass i8
```
Comment (jreback) [inline pandas/src/algos_rank_helper.pxi.in]: remove this comment

File: pandas/src/algos_rank_helper.pxi.in [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+        {{endif}}
+
+        count += 1.0
+
+        {{if name == 'generic'}}
+        if (i == n - 1 or
+            are_diff(util.get_value_at(sorted_data, i + 1), val)):
+        {{elif name == 'float64'}}
+        if i == n - 1 or sorted_data[i + 1] != val:
+        {{else}}
```
Comment (jreback) [inline pandas/src/algos_rank_helper.pxi.in]: this could be nogil

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/src/algos_rank_helper.pxi.in (hunk 5 above)
Comment (jreback) [inline pandas/src/algos_rank_helper.pxi.in]: nvm, this is fine.

File: pandas/src/algos_rank_helper.pxi.in (hunk 6 above)
Comment (gfyoung) [inline pandas/src/algos_rank_helper.pxi.in]: Done.

File: pandas/src/algos_rank_helper.pxi.in (hunk 7 above)
Comment (gfyoung) [inline pandas/src/algos_rank_helper.pxi.in]: Only for non-`object` but agreed.  Done.

**Comment by gfyoung** [comment]
@jreback : Addressed all comments, and both Appveyor and Travis are passing.  Ready to merge if there are no other concerns.

File: pandas/core/algorithms.py [hunk 8]
```diff
@@ -366,6 +366,9 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
     if isinstance(values, Index):
         uniques = values._shallow_copy(uniques, name=None)
     elif isinstance(values, Series):
+        # TODO: This constructor is bugged for uint's, especially
+        # np.uint64 due to overflow. Test this for uint behavior
+        # once constructor has been fixed.
```
Comment (jreback) [inline pandas/core/algorithms.py]: meaning where is the issue?

File: pandas/src/algos_rank_helper.pxi.in [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+        if not ascending:
+            tiebreak = TIEBREAK_FIRST_DESCENDING
+    else:
+        _as = values.argsort(1)
+    {{endif}}
+
+    if not ascending:
+        _as = _as[:, ::-1]
+
+    {{if name == 'generic'}}
```
Comment (jreback) [inline pandas/src/algos_rank_helper.pxi.in]: maybe change generic -> object so this is not necessary? (in a higher level definition)

File: pandas/src/algos_rank_helper.pxi.in [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+"""
+
+#----------------------------------------------------------------------
+# rank_1d, rank_2d
+#----------------------------------------------------------------------
+
+{{py:
+
+# name ctype pos_nan_value neg_nan_value
+dtypes = [('generic', 'object', 'Infinity()', 'NegInfinity()'),
```
Comment (jreback) [inline pandas/src/algos_rank_helper.pxi.in]: here I mean, I don't think we use generic generically (and instead use object) nomenclature

File: pandas/core/algorithms.py (hunk 8 above)
Comment (gfyoung) [inline pandas/core/algorithms.py]: @jreback : I'm handling that TODO right now in #14937.  I'm literally tackling all of these problems simultaneously.

File: pandas/src/algos_rank_helper.pxi.in (hunk 9 above)
Comment (gfyoung) [inline pandas/src/algos_rank_helper.pxi.in]: Yeah...that makes sense.  I haven't been giving too much thought to changing things, but as Travis and Appveyor are happy right now, it couldn't hurt to try.

File: pandas/src/algos_rank_helper.pxi.in (hunk 10 above)
Comment (gfyoung) [inline pandas/src/algos_rank_helper.pxi.in]: Yep, done.

File: pandas/core/algorithms.py (hunk 8 above)
Comment (jreback) [inline pandas/core/algorithms.py]: ok sure

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by gfyoung** [comment]
@jreback : Everything is green now.  Ready to merge if there are no other concerns.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**